File: src/core/orchestrator/agi.py
"""
AGI v2 components — reflection, world model, tool registry, collaboration, code evolution,
vector memory. All loaded lazily and silently skipped if unavailable.
"""

from typing import Any, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from rich.console import Console


class AGIComponents:
    """
    Container and initializer for optional AGI v2 subsystems.
    Each component is loaded lazily and silently ignored on ImportError.
    """

    def __init__(self, llm_client: Optional[Any] = None) -> None:
        self.reflection: Optional[Any] = None
        self.world_model: Optional[Any] = None
        self.tool_registry: Optional[Any] = None
        self.collaboration: Optional[Any] = None
        self.code_evolution: Optional[Any] = None
        self.vector_memory: Optional[Any] = None

        self._load(llm_client)

    def _load(self, llm_client: Optional[Any]) -> None:
        """Attempt to load each AGI subsystem; ignore failures."""
        try:
            from ..reflection import ReflectionEngine
            self.reflection = ReflectionEngine(llm_client=llm_client)
        except Exception:
            pass

        try:
            from ..world_model import WorldModel
            self.world_model = WorldModel(llm_client=llm_client)
        except Exception:
            pass

        try:
            from ..tool_registry import ToolRegistry
            self.tool_registry = ToolRegistry()
        except Exception:
            pass

        try:
            from ..collaboration import CollaborationProtocol
            self.collaboration = CollaborationProtocol(llm_client=llm_client)
        except Exception:
            pass

        try:
            from ..code_evolution import CodeEvolutionEngine
            self.code_evolution = CodeEvolutionEngine(llm_client=llm_client)
        except Exception:
            pass

        try:
            from ..vector_memory_store import VectorMemoryStore
            self.vector_memory = VectorMemoryStore()
        except Exception:
            pass

    # ------------------------------------------------------------------
    # Pre-execution hints (called by runner before workflow starts)
    # ------------------------------------------------------------------

    def print_pre_execution_hints(self, goal: str, console: "Console") -> Any:
        """Print AGI hints to console; return world snapshot (may be None)."""
        world_before = None

        if self.world_model:
            try:
                world_before = self.world_model.snapshot()
                console.print("[dim]🌍 World snapshot captured[/dim]")
            except Exception:
                pass

        if self.reflection:
            try:
                hint = self.reflection.get_strategy_suggestion(goal)
                if hint:
                    console.print(f"[dim]🪞 Strategy hint: {hint[:60]}[/dim]")
            except Exception:
                pass

        if self.tool_registry:
            try:
                suggested = self.tool_registry.suggest_tool(goal)
                if suggested:
                    console.print(
                        f"[dim]🔧 Tool available: {suggested.name} — "
                        f"{suggested.description[:40]}[/dim]"
                    )
            except Exception:
                pass

        if self.world_model:
            try:
                prediction = self.world_model.predict_side_effects(goal)
                if prediction.risk_level == "high":
                    console.print(
                        f"[bold yellow]⚠️  High-risk action detected: "
                        f"{'; '.join(prediction.warnings[:2])}[/bold yellow]"
                    )
            except Exception:
                pass

        if self.vector_memory:
            try:
                from ..vector_memory_store import VectorMemoryStore
                vec = VectorMemoryStore.text_to_hash_vector(goal)
                results = self.vector_memory.search("goal_history", vec, top_k=1)
                if results and results[0].get("score", 0) > 0.85:
                    prev = results[0].get("payload", {}).get("goal", "")
                    console.print(
                        f"[dim]🧠 Similar past goal: {prev[:50]} "
                        f"({results[0]['score']:.0%} match)[/dim]"
                    )
            except Exception:
                pass

        if self.collaboration:
            try:
                roles = self.collaboration.assign_roles(goal)
                if roles:
                    assigned = [f"{r['agent']}: {r['role']}" for r in roles[:2]]
                    console.print(f"[dim]🤝 Agents assigned: {', '.join(assigned)}[/dim]")
            except Exception:
                pass

        return world_before

    # ------------------------------------------------------------------
    # Post-execution pipeline (reflection, world diff, evolution, memory)
    # ------------------------------------------------------------------

    def run_post_execution(
        self,
        goal: str,
        status: str,
        duration_ms: float,
        world_before: Optional[Any],
        errors: List[str],
        console: "Console",
    ) -> None:
        """Run full post-execution AGI pipeline."""
        if self.reflection:
            try:
                reflection = self.reflection.reflect(
                    goal=goal,
                    status=status,
                    duration_ms=duration_ms,
                    error=errors[0] if errors else "",
                )
                if reflection.lesson_learned:
                    console.print(
                        f"\n[dim]🪞 Reflection: {reflection.lesson_learned[:80]}[/dim]"
                    )
                if reflection.strategy_change:
                    console.print(
                        f"[dim]🪞 Strategy change: {reflection.strategy_change[:60]}[/dim]"
                    )
            except Exception:
                pass

        if self.world_model and world_before:
            try:
                world_after = self.world_model.snapshot()
                world_diff = self.world_model.diff(world_before, world_after)
                diff_summary = world_diff.summary()
                if diff_summary and diff_summary != "No changes detected":
                    console.print(
                        f"\n[dim]🌍 World changes: {diff_summary[:100]}[/dim]"
                    )
            except Exception:
                pass

        if self.code_evolution:
            try:
                _journal = self.code_evolution.get_journal(limit=1)
                stats = self.code_evolution.get_stats()
                if stats.get("total_attempts", 0) > 0:
                    console.print(
                        f"[dim]🧬 Evolution: {stats['total_attempts']} attempts, "
                        f"{stats.get('success_rate', 0):.0%} success rate[/dim]"
                    )
            except Exception:
                pass

        if self.vector_memory:
            try:
                import hashlib
                from ..vector_memory_store import VectorMemoryStore
                vec = VectorMemoryStore.text_to_hash_vector(goal)
                goal_id = hashlib.md5(goal.encode()).hexdigest()[:12]
                self.vector_memory.get_or_create_collection("goal_history")
                self.vector_memory.upsert(
                    collection="goal_history",
                    id=goal_id,
                    vector=vec,
                    payload={
                        "goal": goal,
                        "status": status,
                        "duration_ms": duration_ms,
                        "errors": errors[:3] if errors else [],
                    },
                )
            except Exception:
                pass

        if self.collaboration:
            try:
                self.collaboration.submit_review(
                    reviewer="orchestrator",
                    target=goal[:30],
                    approved=status == "success",
                    feedback=[errors[0]] if errors else ["Completed successfully"],
                )
            except Exception:
                pass


__all__ = ["AGIComponents"]

File: .claude/skills/pev-engine/src/agi.ts
/**
 * AGI Components — optional subsystems loaded lazily.
 *
 * Each component (reflection, world model, tool registry, collaboration,
 * code evolution, vector memory) is loaded on demand and silently skipped
 * if unavailable. This keeps the core engine lightweight.
 *
 * In TypeScript/ClaudeKit, these are stub interfaces that can be implemented
 * by consumers. The engine works without any of them.
 */

import type { RecipeStep, ExecutionResult } from "./types";

// ---------------------------------------------------------------------------
// Stub interfaces (implement these in your application)
// ---------------------------------------------------------------------------

export interface ReflectionEngine {
  reflect(params: {
    goal: string;
    status: string;
    durationMs: number;
    error: string;
  }): ReflectionResult;
  getStrategySuggestion(command: string): string | null;
}

export interface ReflectionResult {
  lessonLearned?: string;
  strategyChange?: string;
}

export interface WorldModel {
  snapshot(): WorldSnapshot;
  diff(before: WorldSnapshot, after: WorldSnapshot): WorldDiff;
  predictSideEffects(goal: string): RiskPrediction;
}

export interface WorldSnapshot {
  timestamp: number;
  data: Record<string, unknown>;
}

export interface WorldDiff {
  summary(): string;
}

export interface RiskPrediction {
  riskLevel: "low" | "medium" | "high";
  warnings: string[];
}

export interface ToolRegistry {
  suggestTool(goal: string): ToolSuggestion | null;
  execute(name: string, args: Record<string, unknown>): ToolResult;
}

export interface ToolSuggestion {
  name: string;
  description: string;
}

export interface ToolResult {
  success: boolean;
  output: string;
  error?: string;
}

export interface CollaborationProtocol {
  assignRoles(goal: string): RoleAssignment[];
  submitReview(params: {
    reviewer: string;
    target: string;
    approved: boolean;
    feedback: string[];
  }): void;
}

export interface RoleAssignment {
  agent: string;
  role: string;
}

export interface CodeEvolutionEngine {
  getJournal(limit: number): CodeChange[];
  getStats(): Record<string, number>;
}

export interface CodeChange {
  id: string;
  description: string;
  timestamp: number;
}

export interface VectorMemoryStore {
  search(collection: string, vector: number[], topK: number): MemoryResult[];
  upsert(params: {
    collection: string;
    id: string;
    vector: number[];
    payload: Record<string, unknown>;
  }): void;
  getOrCreateCollection(name: string): void;
}

export interface MemoryResult {
  score: number;
  payload: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// AGIComponents container
// ---------------------------------------------------------------------------

export type AGIComponentMap = {
  reflection: ReflectionEngine | null;
  worldModel: WorldModel | null;
  toolRegistry: ToolRegistry | null;
  collaboration: CollaborationProtocol | null;
  codeEvolution: CodeEvolutionEngine | null;
  vectorMemory: VectorMemoryStore | null;
};

export class AGIComponents implements AGIComponentMap {
  reflection: ReflectionEngine | null = null;
  worldModel: WorldModel | null = null;
  toolRegistry: ToolRegistry | null = null;
  collaboration: CollaborationProtocol | null = null;
  codeEvolution: CodeEvolutionEngine | null = null;
  vectorMemory: VectorMemoryStore | null = null;

  constructor(_components?: Partial<AGIComponentMap>) {
    // Components are set externally by the orchestrator.
    // No lazy loading in TS — consumers inject what they have.
    if (_components) {
      Object.assign(this, _components);
    }
  }

  // -------------------------------------------------------------------------
  // Pre-execution hints
  // -------------------------------------------------------------------------

  printPreExecutionHints(
    _goal: string,
    _log: (msg: string) => void,
  ): WorldSnapshot | null {
    let worldBefore: WorldSnapshot | null = null;

    if (this.worldModel) {
      try {
        worldBefore = this.worldModel.snapshot();
        _log("🌍 World snapshot captured");
      } catch {
        // silently skip
      }
    }

    if (this.reflection) {
      try {
        const hint = this.reflection.getStrategySuggestion(_goal);
        if (hint) _log(`🪞 Strategy hint: ${hint.slice(0, 60)}`);
      } catch {
        // silently skip
      }
    }

    if (this.toolRegistry) {
      try {
        const suggested = this.toolRegistry.suggestTool(_goal);
        if (suggested) {
          _log(
            `🔧 Tool available: ${suggested.name} — ${suggested.description.slice(0, 40)}`,
          );
        }
      } catch {
        // silently skip
      }
    }

    if (this.worldModel) {
      try {
        const prediction = this.worldModel.predictSideEffects(_goal);
        if (prediction.riskLevel === "high") {
          _log(
            `⚠️ High-risk action detected: ${prediction.warnings.slice(0, 2).join("; ")}`,
          );
        }
      } catch {
        // silently skip
      }
    }

    if (this.collaboration) {
      try {
        const roles = this.collaboration.assignRoles(_goal);
        if (roles.length > 0) {
          const assigned = roles
            .slice(0, 2)
            .map((r) => `${r.agent}: ${r.role}`)
            .join(", ");
          _log(`🤝 Agents assigned: ${assigned}`);
        }
      } catch {
        // silently skip
      }
    }

    return worldBefore;
  }

  // -------------------------------------------------------------------------
  // Post-execution pipeline
  // -------------------------------------------------------------------------

  runPostExecution(params: {
    goal: string;
    status: string;
    durationMs: number;
    worldBefore: WorldSnapshot | null;
    errors: string[];
    log: (msg: string) => void;
  }): void {
    const { goal, status, durationMs, worldBefore, errors, log } = params;

    // Reflection
    if (this.reflection) {
      try {
        const reflection = this.reflection.reflect({
          goal,
          status,
          durationMs,
          error: errors[0] ?? "",
        });
        if (reflection.lessonLearned) {
          log(`🪞 Reflection: ${reflection.lessonLearned.slice(0, 80)}`);
        }
        if (reflection.strategyChange) {
          log(`🪞 Strategy change: ${reflection.strategyChange.slice(0, 60)}`);
        }
      } catch {
        // silently skip
      }
    }

    // World diff
    if (this.worldModel && worldBefore) {
      try {
        const worldAfter = this.worldModel.snapshot();
        const worldDiff = this.worldModel.diff(worldBefore, worldAfter);
        const diffSummary = worldDiff.summary();
        if (diffSummary && diffSummary !== "No changes detected") {
          log(`🌍 World changes: ${diffSummary.slice(0, 100)}`);
        }
      } catch {
        // silently skip
      }
    }

    // Code evolution stats
    if (this.codeEvolution) {
      try {
        const stats = this.codeEvolution.getStats();
        const attempts = stats.total_attempts ?? 0;
        if (attempts > 0) {
          const rate = stats.success_rate ?? 0;
          log(`🧬 Evolution: ${attempts} attempts, ${(rate * 100).toFixed(0)}% success rate`);
        }
      } catch {
        // silently skip
      }
    }

    // Vector memory
    if (this.vectorMemory) {
      try {
        const vec = this.hashGoal(goal);
        const goalId = this.hashId(goal);
        this.vectorMemory.getOrCreateCollection("goal_history");
        this.vectorMemory.upsert({
          collection: "goal_history",
          id: goalId,
          vector: vec,
          payload: {
            goal,
            status,
            duration_ms: durationMs,
            errors: errors.slice(0, 3),
          },
        });
      } catch {
        // silently skip
      }
    }

    // Collaboration review
    if (this.collaboration) {
      try {
        this.collaboration.submitReview({
          reviewer: "orchestrator",
          target: goal.slice(0, 30),
          approved: status === "success",
          feedback: errors.length > 0 ? [errors[0]] : ["Completed successfully"],
        });
      } catch {
        // silently skip
      }
    }
  }

  // -------------------------------------------------------------------------
  // Utility
  // -------------------------------------------------------------------------

  private hashGoal(goal: string): number[] {
    // Simple hash → vector placeholder (replace with real embedding)
    let hash = 0;
    for (let i = 0; i < goal.length; i++) {
      hash = ((hash << 5) - hash + goal.charCodeAt(i)) | 0;
    }
    // Produce a 16-dim vector from the hash
    const vec: number[] = [];
    for (let i = 0; i < 16; i++) {
      vec.push(((hash >> i) & 0xff) / 255);
    }
    return vec;
  }

  private hashId(goal: string): string {
    // Simple hash-based ID (use crypto.subtle in production)
    let hash = 0;
    for (let i = 0; i < goal.length; i++) {
      hash = ((hash << 5) - hash + goal.charCodeAt(i)) | 0;
    }
    return Math.abs(hash).toString(16).slice(0, 12);
  }
}
